ntbs-service/DataMigration/NotificationImportService.cs

```diff
@@ -250,5 +255,57 @@ private async Task<List<string>> FilterOutImportedIdsAsync(PerformContext contex
             }
             return legacyIdsToImport;
         }
+        
+        private async Task MigrateOrUpdateCaseManager(Notification notification,
+            PerformContext context,
+            string requestId)
+        {
+            var legacyNotificationCaseManager =
+                (await _migrationRepository.GetNotificationsById(new List<string> {notification.LegacyId}))
+                .Select(n => n.CaseManager)
+                .SingleOrDefault();
+            if (legacyNotificationCaseManager == null)
+            {
+                return;
+            }
+            var existingCaseManager = await _userRepository.GetUserByUsername(legacyNotificationCaseManager);
+            var ntbsNotificationCaseManager = new User();
+            if (existingCaseManager != null)
+            {
+                ntbsNotificationCaseManager = existingCaseManager;
+            }
+            else
+            {
+                ntbsNotificationCaseManager.IsActive = false;
+                ntbsNotificationCaseManager.Username = legacyNotificationCaseManager;
+            }
+
+            var legacyUser = (await _migrationRepository.GetLegacyUserByUsername(legacyNotificationCaseManager)).First();
+            ntbsNotificationCaseManager.GivenName = legacyUser.GivenName;
+            ntbsNotificationCaseManager.FamilyName = legacyUser.FamilyName;
+            ntbsNotificationCaseManager.DisplayName = ntbsNotificationCaseManager.GivenName + " " + ntbsNotificationCaseManager.FamilyName;
+
+            var existingCaseManagerTbServices = await _referenceDataRepository.GetCaseManagerTbServicesByUsernameAsync
+                (ntbsNotificationCaseManager.Username);
+            var ntbsNotificationCaseManagerTbServices = existingCaseManagerTbServices.Select(e => e.TbService).ToList();
+            if (!ntbsNotificationCaseManager.IsActive && existingCaseManagerTbServices.All(cmtb =>
+                cmtb.TbServiceCode != notification.HospitalDetails.TBServiceCode))
+            {
+                var legacyUserHospitals =
+                    (await _migrationRepository.GetLegacyUserHospitalsByUsername(legacyNotificationCaseManager))
+                    .Where(h => h.HospitalId != null).Select(h => h.HospitalId).OfType<Guid>();
+                var legacyUserTbServices =
+                    await _referenceDataRepository.GetTbServicesFromHospitalIdsAsync(legacyUserHospitals);
+                foreach (var tbService in legacyUserTbServices.Where(tb =>
+                    tb.Code == notification.HospitalDetails.TBServiceCode))
```

Comment:
We're using a `foreach` here, but it looks like the where clause should only ever get one result?

If that's the case then I think it would be clearer to have a `SingleOrDefault` call followed by a null check.